File: android/src/main/java/com/reactnativereadium/utils/MetadataNormalizer.kt
package com.reactnativereadium.utils

import com.facebook.react.bridge.*
import org.readium.r2.shared.publication.Metadata
import org.readium.r2.shared.publication.LocalizedString
import org.readium.r2.shared.publication.Contributor
import org.readium.r2.shared.publication.Subject

/**
 * Normalizes metadata to ensure consistent structure per RWPM spec.
 *
 * This normalizer works directly with Readium's native Kotlin types (Metadata, LocalizedString,
 * Contributor, Subject) and converts them to normalized, spec-compliant structures.
 *
 * This approach is:
 * - More efficient (no intermediate WritableMap conversion)
 * - More type-safe (working with actual Kotlin types)
 * - Based on the stable RWPM specification
 *
 * https://readium.org/webpub-manifest/
 */
object MetadataNormalizer {

    /**
     * Normalizes a Metadata object to a consistent React Native-compatible format.
     * Works directly with Readium native types for efficiency and type safety.
     */
    fun normalize(metadata: Metadata): WritableMap {
        // Build typed NormalizedMetadata object from Readium's native types
        val normalized = NormalizedMetadata(
            // Required fields (provide default empty string if null)
            title = normalizeLocalizedString(metadata.localizedTitle) ?: "",

            // Optional localized string fields
            sortAs = normalizeLocalizedString(metadata.localizedSortAs),
            subtitle = normalizeLocalizedString(metadata.localizedSubtitle),

            // Simple string fields
            identifier = metadata.identifier,
            description = metadata.description,
            readingProgression = metadata.readingProgression?.name?.lowercase(),

            // Date fields (convert Instant to ISO string)
            modified = metadata.modified?.toString(),
            published = metadata.published?.toString(),

            // Layout is not directly on Metadata - would need Publication.metadata.presentation
            // For now, pass null and extract from JSON if needed
            layout = null,

            // Array fields
            language = if (metadata.languages.isNotEmpty()) metadata.languages else null,
            conformsTo = metadata.conformsTo.map { it.uri }.takeIf { it.isNotEmpty() },

            // Numeric fields
            duration = metadata.duration,
            numberOfPages = metadata.numberOfPages,

            // Normalize contributors (all types)
            author = normalizeContributors(metadata.authors),
            translator = normalizeContributors(metadata.translators),
            editor = normalizeContributors(metadata.editors),
            artist = normalizeContributors(metadata.artists),
            illustrator = normalizeContributors(metadata.illustrators),
            letterer = normalizeContributors(metadata.letterers),
            penciler = normalizeContributors(metadata.pencilers),
            colorist = normalizeContributors(metadata.colorists),
            inker = normalizeContributors(metadata.inkers),
            narrator = normalizeContributors(metadata.narrators),
            contributor = normalizeContributors(metadata.contributors),
            publisher = normalizeContributors(metadata.publishers),
            imprint = normalizeContributors(metadata.imprints),

            // Normalize subjects
            subject = normalizeSubjects(metadata.subjects),

            // Complex objects - these need special handling
            // For now, convert via JSON if they exist
            accessibility = metadata.accessibility?.let { convertAccessibilityToWritableMap(it) },
            belongsTo = if (metadata.belongsTo.isNotEmpty()) convertBelongsToToWritableMap(metadata.belongsTo) else null
        )

        // Convert to WritableMap for React Native
        return normalized.toWritableMap()
    }

    /**
     * Normalizes a LocalizedString per RWPM spec.
     *
     * Per RWPM spec, we extract a single string from the localized string using priority:
     * 1. null key (undefined/undetermined language)
     * 2. "en" (English fallback)
     * 3. First available translation
     *
     * https://readium.org/webpub-manifest/contexts/default/
     */
    private fun normalizeLocalizedString(value: LocalizedString?): String? {
        if (value == null) return null

        // Priority: null key (undefined) -> "en" -> first available
        return value.translations[null]?.string
            ?: value.translations["en"]?.string
            ?: value.translations.values.firstOrNull()?.string
    }

    /**
     * Normalizes contributors per RWPM spec.
     *
     * Converts Readium's Contributor objects to our normalized format.
     * https://readium.org/webpub-manifest/schema/contributor-object.schema.json
     */
    private fun normalizeContributors(contributors: List<Contributor>): List<NormalizedContributor>? {
        if (contributors.isEmpty()) return null

        return contributors.mapNotNull { contributor ->
            val name = normalizeLocalizedString(contributor.localizedName) ?: return@mapNotNull null

            NormalizedContributor(
                name = name,
                sortAs = normalizeLocalizedString(contributor.localizedSortAs),
                identifier = contributor.identifier,
                role = contributor.roles.firstOrNull(),
                position = contributor.position?.toInt()
            )
        }.takeIf { it.isNotEmpty() }
    }

    /**
     * Normalizes subjects per RWPM spec.
     *
     * Converts Readium's Subject objects to our normalized format.
     * https://readium.org/webpub-manifest/schema/subject.schema.json
     */
    private fun normalizeSubjects(subjects: List<Subject>): List<NormalizedSubject>? {
        if (subjects.isEmpty()) return null

        return subjects.mapNotNull { subject ->
            val name = normalizeLocalizedString(subject.localizedName) ?: return@mapNotNull null

            NormalizedSubject(
                name = name,
                sortAs = normalizeLocalizedString(subject.localizedSortAs),
                code = subject.code,
                scheme = subject.scheme
            )
        }.takeIf { it.isNotEmpty() }
    }

    /**
     * Converts Readium's Accessibility object to WritableMap.
     * This is a complex nested structure that we pass through as-is.
     */
    private fun convertAccessibilityToWritableMap(accessibility: org.readium.r2.shared.publication.Accessibility): WritableMap? {
        return try {
            // Use Readium's built-in JSON serialization
            val json = accessibility.toJSON()
            json.toWritableMap()
        } catch (e: Exception) {
            null
        }
    }

    /**
     * Converts Readium's belongsTo map to WritableMap.
     * This is a complex nested structure that we pass through as-is.
     */
    private fun convertBelongsToToWritableMap(belongsTo: Map<String, List<org.readium.r2.shared.publication.Collection>>): WritableMap? {
        return try {
            Arguments.createMap().apply {
                belongsTo.forEach { (key, collections) ->
                    val array = Arguments.createArray().apply {
                        collections.forEach { collection ->
                            collection.toJSON().toWritableMap()?.let { pushMap(it) }
                        }
                    }
                    putArray(key, array)
                }
            }
        } catch (e: Exception) {
            null
        }
    }
}

File: web/utils/metadataNormalizer.ts
/**
 * Normalizes metadata to ensure consistent structure per RWPM spec.
 *
 * This normalizer is library-agnostic and based on the stable RWPM specification,
 * ensuring it works regardless of how the underlying Readium library serializes data.
 *
 * https://readium.org/webpub-manifest/
 */

/**
 * Normalizes a localized string per RWPM spec.
 *
 * Per RWPM spec, a localized string can be:
 * 1. Plain string: "Moby-Dick"
 * 2. Language map: {"en": "Moby-Dick", "fr": "Moby Dick"}
 *
 * Priority for language selection:
 * 1. "undefined" or "und" (default/undetermined language)
 * 2. "en" (English fallback)
 * 3. First available language
 *
 * https://readium.org/webpub-manifest/contexts/default/
 */
export function normalizeLocalizedString(value: any): string {
  if (!value) return '';

  // Format 1: Plain string (spec-compliant shorthand)
  if (typeof value === 'string') return value;

  // Format 2: Language map (spec-compliant full form)
  if (typeof value === 'object') {
    // Check if it's a LocalizedString instance with translations property
    if (value.translations) {
      const translations = value.translations;
      return translations['undefined'] ||
             translations['und'] ||
             translations['en'] ||
             Object.values(translations)[0] as string || '';
    }

    // Check if it's already a plain language map object
    return value['undefined'] ||
           value['und'] ||
           value['en'] ||
           Object.values(value)[0] as string || '';
  }

  return '';
}

/**
 * Normalizes contributors per RWPM spec.
 *
 * Per RWPM spec, a contributor can be:
 * 1. Plain string: "Herman Melville"
 * 2. Object: {"name": "Herman Melville", "sortAs": "Melville, Herman"}
 *
 * The Web library uses a Contributors class with an "items" property.
 *
 * https://readium.org/webpub-manifest/schema/contributor-object.schema.json
 */
export function normalizeContributors(value: any): any[] | undefined {
  if (!value) return undefined;

  // Handle TypeScript library's Contributors class format (has .items property)
  const items = value.items || value;

  if (!Array.isArray(items)) return undefined;

  return items.map((contributor: any) => {
    // Format 1: Plain string
    if (typeof contributor === 'string') {
      return { name: contributor };
    }

    // Format 2: Object with potentially localized fields
    const normalized: any = {
      name: normalizeLocalizedString(contributor.name),
    };

    if (contributor.sortAs) {
      normalized.sortAs = normalizeLocalizedString(contributor.sortAs);
    }
    if (contributor.identifier) {
      normalized.identifier = contributor.identifier;
    }
    if (contributor.role) {
      normalized.role = contributor.role;
    }
    if (contributor.position !== undefined) {
      normalized.position = contributor.position;
    }

    return normalized;
  }).filter(c => c.name); // Remove contributors without names
}

/**
 * Normalizes subjects per RWPM spec.
 *
 * Subject names can be LocalizedStrings.
 * The Web library uses a Subjects class with an "items" property.
 *
 * https://readium.org/webpub-manifest/schema/subject.schema.json
 */
export function normalizeSubjects(value: any): any[] | undefined {
  if (!value) return undefined;

  // Handle Subjects class format (has .items property)
  const items = value.items || value;

  if (!Array.isArray(items)) return undefined;

  return items.map((subject: any) => {
    // Format 1: Plain string
    if (typeof subject === 'string') {
      return { name: subject };
    }

    // Format 2: Object
    const normalized: any = {
      name: normalizeLocalizedString(subject.name),
    };

    if (subject.sortAs) {
      normalized.sortAs = normalizeLocalizedString(subject.sortAs);
    }
    if (subject.code) {
      normalized.code = subject.code;
    }
    if (subject.scheme) {
      normalized.scheme = subject.scheme;
    }

    return normalized;
  }).filter(s => s.name);
}

/**
 * Normalizes the entire metadata object per RWPM spec.
 *
 * This function is library-agnostic and based on the stable RWPM specification,
 * ensuring it works regardless of how the underlying Readium library serializes data.
 *
 * https://readium.org/webpub-manifest/
 */
export function normalizeMetadata(metadata: any): any {
  const normalized: any = {
    // Title is required
    title: normalizeLocalizedString(metadata.title),
  };

  // Optional localized string fields
  if (metadata.subtitle) {
    normalized.subtitle = normalizeLocalizedString(metadata.subtitle);
  }
  if (metadata.sortAs) {
    normalized.sortAs = normalizeLocalizedString(metadata.sortAs);
  }

  // Direct copy fields
  if (metadata.identifier) normalized.identifier = metadata.identifier;
  if (metadata.conformsTo) normalized.conformsTo = metadata.conformsTo;
  if (metadata.description) normalized.description = metadata.description;
  if (metadata.readingProgression) normalized.readingProgression = metadata.readingProgression;
  if (metadata.layout) normalized.layout = metadata.layout;
  if (metadata.duration) normalized.duration = metadata.duration;
  if (metadata.numberOfPages) normalized.numberOfPages = metadata.numberOfPages;

  // Date fields - convert to ISO strings
  if (metadata.modified) {
    normalized.modified = metadata.modified instanceof Date
      ? metadata.modified.toISOString()
      : metadata.modified;
  }
  if (metadata.published) {
    normalized.published = metadata.published instanceof Date
      ? metadata.published.toISOString()
      : metadata.published;
  }

  // Languages (Web uses plural 'languages', our interface uses singular 'language')
  if (metadata.languages) {
    normalized.language = metadata.languages;
  }

  // Normalize all contributor types (Web uses plural names, our interface uses singular)
  if (metadata.authors) normalized.author = normalizeContributors(metadata.authors);
  if (metadata.translators) normalized.translator = normalizeContributors(metadata.translators);
  if (metadata.editors) normalized.editor = normalizeContributors(metadata.editors);
  if (metadata.artists) normalized.artist = normalizeContributors(metadata.artists);
  if (metadata.illustrators) normalized.illustrator = normalizeContributors(metadata.illustrators);
  if (metadata.letterers) normalized.letterer = normalizeContributors(metadata.letterers);
  if (metadata.pencilers) normalized.penciler = normalizeContributors(metadata.pencilers);
  if (metadata.colorists) normalized.colorist = normalizeContributors(metadata.colorists);
  if (metadata.inkers) normalized.inker = normalizeContributors(metadata.inkers);
  if (metadata.narrators) normalized.narrator = normalizeContributors(metadata.narrators);
  if (metadata.contributors) normalized.contributor = normalizeContributors(metadata.contributors);
  if (metadata.publishers) normalized.publisher = normalizeContributors(metadata.publishers);
  if (metadata.imprints) normalized.imprint = normalizeContributors(metadata.imprints);

  // Normalize subjects
  if (metadata.subjects) {
    normalized.subject = normalizeSubjects(metadata.subjects);
  }

  // Copy complex objects as-is
  if (metadata.belongsTo) normalized.belongsTo = metadata.belongsTo;
  if (metadata.accessibility) normalized.accessibility = metadata.accessibility;

  return normalized;
}
